File: Data Structures/Searching and Sorting/MergeSort.py
#Code for Merge Sort in Python
#Rishabh Pathak

def mergeSort(lst):
    if len(lst) > 1:
        mid = len(lst) // 2

        #dividing the list into left and right halves
        left = lst[:mid]
        right = lst[mid:]

        #recursive call for further divisions
        mergeSort(left)
        mergeSort(right)

        #iterators for left and right halves
        i = 0
        j = 0

        #iterator for main list
        k = 0

        while i < len(left) and j < len(right):
            #comparing values from left half with the values from right half and inserting the smaller value in the main list
            if left[i] < right[j]:
                lst[k] = left[i]
                i += 1   #incrementing value of iterator for left half
            else:
                lst[k] = right[j]
                j += 1   #incrementing value of iterator for right half

            k += 1       #incrementing value of iterator for main list

        #for all the remaining values
        while i < len(left):
            lst[k] = left[i]
            i += 1
            k += 1

        while j < len(right):
            lst[k] = right[j]
            j += 1
            k += 1

#an example
myList = [5, 7, 6, 1, 9]
mergeSort(myList)
print(myList) 

#Time complexity: O(nlogn)

File: Data Structures/Searching and Sorting/MergeSort.cpp
//merge sort
#include<stdio.h>
void merge(int a[],int ,int ,int);
void merge_sort(int a[],int ,int);
int main()
{
    int a[10],i,n,j,k;
    printf("Enter the number of elements in the array:");
    scanf("%d", &n);
    for(i=0;i<n;i++)
    {
        printf("A[%d]= ",i);
        scanf("%d",&a[i]);
    }
    merge_sort(a,0,n-1);
    printf("The sorted array is:");
    for(i=0;i<n;i++)
        printf("%d\t",a[i]);
}
void merge_sort(int a[], int beg, int end)
{
    int mid;
    if(beg<end)
    {
        mid=(beg+end)/2;
        merge_sort(a,beg,mid);
        merge_sort(a,mid+1,end);
        merge(a,beg,mid,end);
    }
}
void merge(int a[],int beg, int mid, int end)
{
    int i=beg, j=mid+1, index=beg, temp[10], k;
    while((i<=mid)&&(j<=end))
    {
        if(a[i]<a[j])
        {
            temp[index]=a[i];
            i++;
        }
        else
        {
            temp[index]=a[j];
            j++;
        }
        index++;
    }
    if(i>mid)
    {
        while(j<=end)
        {
            temp[index]=a[j];
            j++;
            index;
        }
    }
    else
    {
        while(i<=mid)
        {
            temp[index]=a[i];
            i++;
            index++;
        }
    }
    for(k=beg;k<index;k++)
    {
        a[k]=temp[k];
    }
}

